packages/components/src/internal/query/api.ts

```diff
@@ -41,49 +41,99 @@ export function invalidateFullQueryDetailsCache(): void {
     queryDetailsCache = {};
 }
 
-function getQueryDetailsCacheKey(schemaQuery: SchemaQuery, containerPath?: string, fk?: string): string {
-    return '' + schemaQuery.getKey(false) + (fk ? '|' + fk : '') + (containerPath ? '|' + containerPath : '');
-}
+const QUERY_CACHE_KEY_SEPARATOR = '|';
+
+function getQueryDetailsCacheKey(
+    schemaQuery: SchemaQuery,
+    containerPath?: string,
+    fk?: string,
+    fields?: string | string[]
+): string {
+    const parts = [schemaQuery.getKey(false) + QUERY_CACHE_KEY_SEPARATOR];
+    if (fk) parts.push(fk);
+    if (containerPath) parts.push(containerPath);
+    if (fields) {
+        let fields_: string[];
+        if (Array.isArray(fields)) {
+            fields_ = Array.from(fields);
+        } else {
+            fields_ = [fields];
+        }
 
-export function invalidateQueryDetailsCache(schemaQuery: SchemaQuery, containerPath?: string, fk?: string): void {
-    const key = getQueryDetailsCacheKey(schemaQuery, containerPath, fk);
-    delete queryDetailsCache[key];
+        fields_.sort().forEach(field => {
+            if (field) {
+                parts.push(field);
+            }
+        });
+    }
 
-    // also call invalidate for the query key without the containerPath
-    if (containerPath) {
-        const keyNoContainerPath = getQueryDetailsCacheKey(schemaQuery, undefined, fk);
-        delete queryDetailsCache[keyNoContainerPath];
+    return parts.join(QUERY_CACHE_KEY_SEPARATOR);
+}
+
+export function invalidateQueryDetailsCache(
+    schemaQuery: SchemaQuery,
+    containerPath?: string,
+    fk?: string,
+    fields?: string | string[],
+    exactKeyMatch = false
+): void {
+    if (exactKeyMatch) {
+        const key = getQueryDetailsCacheKey(schemaQuery, containerPath, fk, fields);
+        delete queryDetailsCache[key];
+    } else {
+        const prefix = getQueryDetailsCacheKey(schemaQuery);
+        Object.keys(queryDetailsCache).forEach(cacheKey => {
+            if (cacheKey.startsWith(prefix)) {
+                delete queryDetailsCache[cacheKey];
+            }
+        });
     }
 }
 
-export interface GetQueryDetailsOptions {
-    containerPath?: string;
-    fk?: string;
+interface GetQueryDetailsBasic
+    extends Omit<Query.GetQueryDetailsOptions, 'method' | 'schemaName' | 'queryName' | 'viewName'> {
     lookup?: QueryLookup;
+}
+
+interface GetQueryDetailsName extends GetQueryDetailsBasic {
     queryName: string;
     schemaName: string;
-    viewName?: string;
+    schemaQuery?: never;
```

Comment:
what does `never` do here? also, curious if it matters if the prop is set as optional when never is used?

Reply:
This is an attempt to allow for better typings while also supporting different types of properties on the interface. The idea here is that callers can provider either `schemaName/queryName` OR `schemaQuery` but not both and that the typings enforce this. The use of `never` indicates that when this type is matched then the properties typed as `never` are not acceptable as part of this type. This is known as "narrowing" to help narrow down a specific type. https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-never-type